crypto/dh/dh_rfc7919.c

```diff
@@ -47,28 +77,48 @@ DH *DH_new_by_nid(int nid)
 int DH_get_nid(const DH *dh)
 {
     int nid;
+    const BIGNUM *p, *q, *g;
+    BIGNUM *qc = NULL;
+
+    FFC_PARAMS_get0_pqg(&dh->params, &p, &q, &g);
```

Comment:
That is fine when DH_new_by_nid() was used for creation.
It could cache the nid (inside the params struct ?), but it will still need to reset it and do a full search in the case of a user setting the p,q,g or on a ASN1 load.. Would it automagically do the search in the set and load? Also the ASN1 load should probably also set q if it needs to.
I am thinking that this API is actually 'Do these domain parameters represent a safe prime group'.